src/client/lod_mapblock.cpp

```diff
@@ -0,0 +1,345 @@
+// Luanti
+// SPDX-License-Identifier: LGPL-2.1-or-later
+// Copyright (C) 2010-2013 celeron55, Perttu Ahola <celeron55@gmail.com>
+
+#include <cmath>
+#include "lod_mapblock.h"
+#include "util/basic_macros.h"
+#include "util/numeric.h"
+#include "util/tracy_wrapper.h"
+#include "mapblock_mesh.h"
+#include "settings.h"
+#include "nodedef.h"
+#include "client/tile.h"
+#include "client/meshgen/collector.h"
+#include "client/renderingengine.h"
+#include "client.h"
+
+#include "profiler.h"
+
+static constexpr u16 quad_indices_02[] = {0, 1, 2, 2, 3, 0};
+static const auto &quad_indices = quad_indices_02;
+
+LodMeshGenerator::LodMeshGenerator(MeshMakeData *input, MeshCollector *output, const bool is_mono_mat):
+	m_data(input),
+	m_collector(output),
+	m_nodedef(m_data->m_nodedef),
+	m_blockpos_nodes(m_data->m_blockpos * MAP_BLOCKSIZE),
+	m_is_mono_mat(is_mono_mat)
+{
+}
+
+void LodMeshGenerator::generateBitsetMesh(const MapNode n, const u8 width,
+                                          const v3s16 seg_start, const video::SColor color_in)
+{
+	const core::vector3df seg_offset(seg_start.X * BS, seg_start.Y * BS, seg_start.Z * BS);
+	const f32 scaled_BS = BS * width;
+
+	core::vector3df vertices[4];
+	video::S3DVertex irr_vertices[4];
+	for (u8 direction = 0; direction < Direction_END; direction++) {
+		TileSpec tile;
+		video::SColor color;
+		if (m_is_mono_mat) {
+			// When generating a mesh with no texture, we have to color the vertices instead.
+			video::SColor c2 = m_nodedef->get(n).average_colors[direction];
+			color = video::SColor(
+				color.getAlpha(),
+				color_in.getRed() * c2.getRed() / 255U,
+				color_in.getGreen() * c2.getGreen() / 255U,
+				color_in.getBlue() * c2.getBlue() / 255U);
+		} else {
+			getNodeTileN(n, m_blockpos_nodes, direction, m_data, tile);
+			color = color_in;
+		}
+
+		const u64 direction_offset = BITSET_MAX_NOPAD2 * direction;
+		for (u8 slice_i = 0; slice_i < BITSET_MAX_NOPAD; slice_i++) {
+			const u64 slice_offset = direction_offset + BITSET_MAX_NOPAD * slice_i;
+			for (u8 u = 0; u < BITSET_MAX_NOPAD; u++) {
+
+				bitset column = m_slices[slice_offset + u];
+				while (column) {
+					s32 v0 = __builtin_ctzll(column);
```

Comment:
to be clear this probably needs to be wrapped for compilers where this builtin is not available